Can't use Deepspeed in Sentence-transformers

I'm trying to train a model using Deepspeed with 8xH200, but I found an error. Here are my setup details.

Lib versions (python 3.10.18)

```
sentence-transformers       5.2.0.dev0
transformers                4.56.2
deepspeed                   0.17.6
accelerate                  1.10.1
```

Setting
```
args = SentenceTransformerTrainingArguments(
        # Required parameter:
        output_dir=f"models/{run_name}",
        # Optional training parameters:
        batch_sampler=BatchSamplers.NO_DUPLICATES,
        num_train_epochs=1,
        per_device_train_batch_size=1,
        per_device_eval_batch_size=1,
        gradient_accumulation_steps=8,  # Simulate larger batch size
        learning_rate=4e-5,
        warmup_ratio=0.1,
        fp16=False,  # Set to False if you get an error that your GPU can't run on FP16
        bf16=True,  # Set to True if you have a GPU that supports BF16
        gradient_checkpointing=True,  # Trade compute for memory
        dataloader_pin_memory=False,  # Reduce memory usage
        remove_unused_columns=False,  # Keep all columns to avoid extra processing
        dataloader_num_workers=0,  # Reduce memory usage
        deepspeed="deepspeed_zero1_config.json",  # Use ZeRO-1 instead of ZeRO-3
        # Optional tracking/debugging parameters:
        # eval_strategy="steps",
        # eval_steps=500,
        save_strategy="steps",
        save_steps=500,
        save_total_limit=2,  # Reduce number of checkpoints to save memory
        logging_steps=20,
        # run_name=run_name,  # Will be used in W&B if `wandb` is installed
    )
```

Error 
```
[rank7]: Traceback (most recent call last):
[rank7]:   File "/mnt/weka/aisg/peerat/sentence-transformers/pre_training_training.py", line 121, in <module>
[rank7]:     main()
[rank7]:   File "/mnt/weka/aisg/peerat/sentence-transformers/pre_training_training.py", line 107, in main
[rank7]:     trainer.train()
[rank7]:   File "/mnt/weka/home/peerat/.conda/envs/embedding/lib/python3.10/site-packages/transformers/trainer.py", line 2328, in train
[rank7]:     return inner_training_loop(
[rank7]:   File "/mnt/weka/home/peerat/.conda/envs/embedding/lib/python3.10/site-packages/transformers/trainer.py", line 2432, in _inner_training_loop
[rank7]:     self.optimizer, self.lr_scheduler = deepspeed_init(self, num_training_steps=max_steps)
[rank7]:   File "/mnt/weka/home/peerat/.conda/envs/embedding/lib/python3.10/site-packages/transformers/integrations/deepspeed.py", line 444, in deepspeed_init
[rank7]:     hf_deepspeed_config.trainer_config_finalize(args, model, num_training_steps)
[rank7]:   File "/mnt/weka/home/peerat/.conda/envs/embedding/lib/python3.10/site-packages/transformers/integrations/deepspeed.py", line 224, in trainer_config_finalize
[rank7]:     if hasattr(model.config, "hidden_size"):
[rank7]:   File "/mnt/weka/home/peerat/.conda/envs/embedding/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1962, in __getattr__
[rank7]:     raise AttributeError(
[rank7]: AttributeError: 'SentenceTransformer' object has no attribute 'config'
```

Full Code
```
def main():
    model = SentenceTransformer(
        "google-bert/bert-base-multilingual-cased",
    ) # google-bert/bert-base-multilingual-cased, /mnt/weka/aisg/peerat/separated_models/gemma3_text_model

    logging.info("Read datasets")
    train_dataset = load_dataset("tomaarsen/ms-marco-n-tuple-scores-mxbai-embed-large-v1-40000-candidates", split="train")

    anchor = []
    positive = []
    negative = []
    for data in train_dataset:
        anchor.append(data["query"])
        positive.append(data["positive"])
        negative.append(data["negative_1"])

    train_dataset = Dataset.from_dict({
                    "anchor": anchor,
                    "positive": positive,
                    "negative": negative
                })
    train_loss = losses.CachedMultipleNegativesRankingLoss(model, mini_batch_size=32)

    run_name = "sealionv4-msmarco-simcse"
    args = SentenceTransformerTrainingArguments(
        # Required parameter:
        output_dir=f"models/{run_name}",
        # Optional training parameters:
        batch_sampler=BatchSamplers.NO_DUPLICATES,
        num_train_epochs=1,
        per_device_train_batch_size=1,
        per_device_eval_batch_size=1,
        gradient_accumulation_steps=8,  # Simulate larger batch size
        learning_rate=4e-5,
        warmup_ratio=0.1,
        fp16=False,  # Set to False if you get an error that your GPU can't run on FP16
        bf16=True,  # Set to True if you have a GPU that supports BF16
        gradient_checkpointing=True,  # Trade compute for memory
        dataloader_pin_memory=False,  # Reduce memory usage
        remove_unused_columns=False,  # Keep all columns to avoid extra processing
        dataloader_num_workers=0,  # Reduce memory usage
        deepspeed="deepspeed_zero1_config.json",  # Use ZeRO-1 instead of ZeRO-3
        save_strategy="steps",
        save_steps=500,
        save_total_limit=2,  # Reduce number of checkpoints to save memory
        logging_steps=20,
        # run_name=run_name,  # Will be used in W&B if `wandb` is installed
    )

    trainer = SentenceTransformerTrainer(
        model=model,
        args=args,
        train_dataset=train_dataset,
        # eval_dataset=eval_dataset,
        loss=train_loss,
        # evaluator=dev_evaluator,
    )
    trainer.train()
```

Is there anything I wrongly set up? Thank you.

Hello!

Thank you for opening this. I've identified this as an issue/oversight in `transformers`, where the model is expected to have a `config` key. You can avoid that code path by not setting `"reduce_bucket_size"` to `"auto"`, but instead to [hidden_dim](https://huggingface.co/google-bert/bert-base-multilingual-cased/blob/main/config.json#L9) * hidden_dim, 589824 for your model of choice here. This follows what "auto" does: https://github.com/huggingface/transformers/blob/52cbc7c868fde27f10e39504fb74ac733cc98b63/src/transformers/integrations/deepspeed.py#L271-L289

Then, you should be able to start training, although I have a few concerns:
1. I believe that CachedMultipleNegativesRankingLoss does not work with deepspeed. It uses a few hacks that seem incompatible.
2. Although `gradient_accumulation_steps` works to get a higher simulated batch size, it doesn't improve performance in the same way that a higher batch size normally does for the (Cached)MultipleNegativesRankingLoss a.k.a. InfoNCE, as `gradient_accumulation_steps` doesn't result in more in-batch negatives.

If you can, please have a look whether this works for you now. I'm still having some issues on my end, sadly:
```
[rank0]: Traceback (most recent call last):
[rank0]:   File "/mnt/c/code/sentence-transformers/demo_3531.py", line 71, in <module>
[rank0]:     main()
[rank0]:   File "/mnt/c/code/sentence-transformers/demo_3531.py", line 67, in main
[rank0]:     trainer.train()
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/transformers/trainer.py", line 2328, in train
[rank0]:     return inner_training_loop(
[rank0]:            ^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/transformers/trainer.py", line 2672, in _inner_training_loop
[rank0]:     tr_loss_step = self.training_step(model, inputs, num_items_in_batch)
[rank0]:                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/transformers/trainer.py", line 4071, in training_step
[rank0]:     self.accelerator.backward(loss, **kwargs)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/accelerate/accelerator.py", line 2351, in backward
[rank0]:     self.deepspeed_engine_wrapped.backward(loss, **kwargs)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/accelerate/utils/deepspeed.py", line 266, in backward
[rank0]:     self.engine.backward(loss, **kwargs)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/utils/nvtx.py", line 20, in wrapped_fn
[rank0]:     ret_val = func(*args, **kwargs)
[rank0]:               ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/engine.py", line 2323, in backward
[rank0]:     self._do_optimizer_backward(loss, retain_graph)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/engine.py", line 2269, in _do_optimizer_backward        
[rank0]:     self.optimizer.backward(loss, retain_graph=retain_graph)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 2246, in backward
[rank0]:     self.loss_scaler.backward(loss.float(), retain_graph=retain_graph)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/fp16/loss_scaler.py", line 65, in backward
[rank0]:     scaled_loss.backward(retain_graph=retain_graph)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/torch/_tensor.py", line 648, in backward
[rank0]:     torch.autograd.backward(
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/torch/autograd/__init__.py", line 353, in backward
[rank0]:     _engine_run_backward(
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/torch/autograd/graph.py", line 824, in _engine_run_backward
[rank0]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/torch/autograd/function.py", line 307, in apply
[rank0]:     return user_fn(self, *args)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/torch/utils/checkpoint.py", line 320, in backward
[rank0]:     torch.autograd.backward(outputs_with_grad, args_with_grad)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/torch/autograd/__init__.py", line 353, in backward
[rank0]:     _engine_run_backward(
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/torch/autograd/graph.py", line 824, in _engine_run_backward
[rank0]:     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 985, in grad_handling_hook 
[rank0]:     self.process_gradients(param, i)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1524, in process_gradients 
[rank0]:     self.reduce_ready_partitions_and_remove_grads(param, i)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1528, in reduce_ready_partitions_and_remove_grads
[rank0]:     self.reduce_independent_p_g_buckets_and_remove_grads(param, i)
[rank0]:   File "/home/tom/miniconda3/envs/transformers/lib/python3.11/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1017, in reduce_independent_p_g_buckets_and_remove_grads
[rank0]:     assert self.params_already_reduced[param_id] == False, \
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]: AssertionError: The parameter 196 has already been reduced.             Gradient computed twice for this partition.             Multiple gradient reduction is currently not supported
```
But I'm unsure about why this is happening. I've successfully used Deepspeed in the past with Sentence Transformers, but that was many versions ago now.

- Tom Aarsen

--- a/sentence_transformers/base/model.py
+++ b/sentence_transformers/base/model.py
@@ -12,7 +12,7 @@ from abc import ABC, abstractmethod
 from collections import OrderedDict
 from multiprocessing import Queue
 from pathlib import Path
-from typing import Any, Literal
+from typing import TYPE_CHECKING, Any, Literal
 
 import numpy as np
 import torch
@@ -40,6 +40,9 @@ from sentence_transformers.util import (
     save_to_hub_args_decorator,
 )
 from sentence_transformers.util.misc import ORIGINAL_TRANSFORMER_MODELS
+
+if TYPE_CHECKING:
+    from transformers import PretrainedConfig
 
 logger = transformers_logging.get_logger(__name__)
 
@@ -1520,6 +1523,21 @@ This pull request has been automatically generated to add {self.__class__.__name
             if isinstance(module, PreTrainedModel):
                 return module
         return None
+
+    @property
+    def config(self) -> PretrainedConfig | None:
+        """The transformers ``PretrainedConfig`` of the underlying model.
+
+        Several integrations (most notably Deepspeed and ``transformers.Trainer``) read
+        ``model.config.hidden_size`` directly. Without this delegation those integrations
+        crash with ``AttributeError: 'SentenceTransformer' object has no attribute 'config'``
+        because :class:`BaseModel` is an ``nn.Sequential`` rather than a ``PreTrainedModel``.
+
+        Returns the config of :attr:`transformers_model`, or ``None`` when no underlying
+        transformers model can be located (e.g. a pure StaticEmbedding-only setup).
+        """
+        underlying = self.transformers_model
+        return getattr(underlying, "config", None)
 
     @property
     def _target_device(self) -> torch.device:

--- a/sentence_transformers/cross_encoder/model.py
+++ b/sentence_transformers/cross_encoder/model.py
@@ -422,13 +422,6 @@ class CrossEncoder(BaseModel, FitMixin):
         return nn.Identity()
 
     @property
-    def config(self) -> PretrainedConfig | None:
-        transformers_model = self.transformers_model
-        if transformers_model is None:
-            return None
-        return transformers_model.config
-
-    @property
     def model(self) -> PreTrainedModel | None:
         return self.transformers_model
 

--- a/tests/base/test_model.py
+++ b/tests/base/test_model.py
@@ -148,6 +148,24 @@ def test_preprocess_skips_modality_check_for_empty_inputs(
     assert isinstance(features, dict)
 
 
+def test_config_delegates_to_underlying_transformers_model(stsb_bert_tiny_model: SentenceTransformer) -> None:
+    """`model.config` should return the underlying transformers PretrainedConfig so
+    integrations like Deepspeed and transformers.Trainer can read `hidden_size` etc.
+    """
+    assert stsb_bert_tiny_model.config is not None
+    assert hasattr(stsb_bert_tiny_model.config, "hidden_size")
+    assert stsb_bert_tiny_model.config.hidden_size == stsb_bert_tiny_model.transformers_model.config.hidden_size
+
+
+def test_config_returns_none_when_no_underlying_transformers_model(
+    static_embedding_model: SentenceTransformer,
+) -> None:
+    """If the model has no underlying transformers PreTrainedModel (e.g. StaticEmbedding-only),
+    `model.config` returns `None` instead of raising.
+    """
+    assert static_embedding_model.config is None
+
+
 def test_sentence_transformer(stsb_bert_tiny_model: SentenceTransformer) -> None:
     assert stsb_bert_tiny_model.supports("text")
     assert not stsb_bert_tiny_model.supports("image")

